lightningd/peer_control.c

```diff
@@ -923,6 +924,689 @@ static void json_add_channel(struct lightningd *ld,
+	else if (streq(name, "message"))
+		json_add_string(response, alias,
+				state_change->message);
+	else
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "unknown field");
+	return NULL;
+}
+
+struct command_result *json_add_channel_htlc_field(struct json_stream *response,
+						   struct command *cmd,
+						   const struct channel *channel,
+                                                   const void *hinout, bool in,
+                                                   const struct graphql_selection *sel);
+
+struct command_result *json_add_channel_htlc_field(struct json_stream *response,
+						   struct command *cmd,
+						   const struct channel *channel,
+						   const void *hinout, bool in,
+						   const struct graphql_selection *sel)
+{
+	const char *name, *alias;
+
+	if (!sel->field || sel->frag_spread || sel->inline_frag)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "fragments not supported");
+	if (!sel->field->name || sel->field->name->token_type != 'a' ||
+	    !sel->field->name->token_string)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "invalid field");
+	name = alias = sel->field->name->token_string;
+	if (sel->field->alias && sel->field->alias->name &&
+	    sel->field->alias->name->token_type == 'a' &&
+	    sel->field->alias->name->token_string)
+		alias = sel->field->alias->name->token_string;
+
+	/* only one of these will be valid, based on "in" argument */
+	/* use only the correct one */
+	const struct htlc_in *hin = (const struct htlc_in *)hinout;
+	const struct htlc_out *hout = (const struct htlc_out *)hinout;
+
+	if (streq(name, "direction"))
+		json_add_string(response, alias, in? "in":"out");
+	else if (streq(name, "id"))
+		json_add_u64(response, alias, in? hin->key.id: hout->key.id);
+	else if (streq(name, "msatoshi"))
+		json_add_u64(response, alias,
+			     in? hin->msat.millisatoshis: hout->msat.millisatoshis);
+	else if (streq(name, "amount_msat"))
+		json_add_amount_msat_only(response, alias, in? hin->msat: hout->msat);
+	else if (streq(name, "expiry")) {
+		if (in)
+			json_add_u32(response, alias, hin->cltv_expiry);
+		else
+			json_add_u64(response, alias, hout->cltv_expiry);
+	} else if (streq(name, "payment_hash"))
+		json_add_sha256(response, alias,
+				in? &hin->payment_hash: &hout->payment_hash);
+	else if (streq(name, "state"))
+		json_add_string(response, alias,
+				htlc_state_name(in? hin->hstate: hout->hstate));
+	else if (streq(name, "local_trimmed")) {
+		u32 local_feerate = get_feerate(channel->fee_states,
+						channel->opener, LOCAL);
+		json_add_bool(response, alias,
+			htlc_is_trimmed(in?REMOTE:LOCAL, in?hin->msat:hout->msat,
+					local_feerate,
+					channel->our_config.dust_limit, LOCAL,
+					channel_has(channel, OPT_ANCHOR_OUTPUTS))?
+			true : false);
+	} else if (streq(name, "status")) {
+		if (in) {
+			if (hin->status != NULL)
+				json_add_string(response, alias, hin->status);
+			else
+				json_add_null(response, alias);
```

Comment:
No, we don't do this in the old code. This came about because with GraphQL, if the user specifies a field, then that field must appear in the result. (i.e. section 2.5 Fields ....All GraphQL operations must specify their selections down to fields which return scalar values to ensure an unambiguously shaped response....)